### ISSUE TITLE
Question: What if i need two instances of entity context in a controller ?

### ISSUE THREAD

**Issue by TrabacchinLuigi** [aspnet/DependencyInjection#352]
i have investigated many solutions for this *particular need.

But I'd like to receive advice from this community since i wasn't able to get any answer on stack overflow, where [i originally posted my question](http://stackoverflow.com/questions/34787406/having-more-than-one-entity-context), since my concern are about test-ability, injection, the way entity framework is "integrated" in asp.net i thought this was the best place to ask for advice...

*I'm pretty sure many people will use this the same way i'm doing so i guess it can help arguing about ...

**Comment by Eilon** [comment]
@lodejard do you have any thoughts on this?

**Comment by lodejard** [comment]
Hi @TrabacchinLuigi, can you push a tiny sample web app to github showing a small example of the code you'd expect to write which doesn't work? 

I think this might be less of a DI question and more of a DbContext one... @divega, does this situation sound familiar?

**Comment by TrabacchinLuigi** [comment]
I can prepare some samples,  but it will take me some days. 
What is not working actually is the serialization, because of circular references, I do know how to serialize them the point is I don't want to and since I did not any query into the controller function to fetch more data than the one I wanted it is confusing so I guess having more dB contexts would be nice, but how?

**Comment by divega** [comment]
@TrabacchinLuigi Trying to respond first to your SO question and various solution attempts: you are right to think that in some situations setting `context.Database.ChangeTracker.QueryTrackingBehavior = Microsoft.Data.Entity.QueryTrackingBehavior.NoTracking` or just using `AsNoTracking()` on individual queries can help, but other times you'll need to have more than one instance of the DbContext created.

You normally use the `AddDbContext<TContext>()` method you in startup to make sure an instance of your context type is created per request and the right DbContextOptions and service provider get set on it. When you need to deviate from this pattern you have a few options, e.g.:

1. Include a constructor in your derived DbContext class that takes an IServiceProvider and passes it to the base constructor. Make sure your controller takes IServiceProvider. Once you do this you should be able to create DbContext manually with something like this:

 ``` C#
    using( var context1 = new MyDbContext(serviceProvider),
           var context2 = new MyDbContext(serviceProvider))
       ...
 ```
 If you are still using `AddDbContext<MyDbContext>(options => ...)` in startup it should pull those options automatically from the service provider. But you can also choose to either include the `DbContextOptions` as a parameter in the constructor or override the `OnConfiguring()` method for that.

2. To avoid having to change the constructor signatures on your derived DbContext type you can take advantage of the DbContextActivator class (it is our Internal namespace), e.g.:

``` C#
   using(var context1 = DbContextActivator.CreateInstance<MyDbContext>(serviceProvider),
             var context2 = DbContextActivator.CreateInstance<MyDbContext>(serviceProvider)
      ...
```
I am giving you examples that create two separate DbContexts in a using block, but you should also be able to mix those with the regular "per-request" DbContext you would get injected in the controller's constructor.

**Comment by TrabacchinLuigi** [comment]
The only problem in changing the constructor signature in my derived type is that identity framework does not expose an constructor who take IServiceProvider, so, even if it is very easy to re-implement that, i guess many people would not want to, in case it will change in the future...
P.s. are the examples still needed ?

**Comment by Eilon** [comment]
@divega any further thoughts on this?

**Comment by divega** [comment]
Yes, two more thoughts:

- We added all the `IdentityDbContext` constructors that were missing in https://github.com/aspnet/Identity/commit/ada5756c874bec41b19b492e102c31930a09bd5d a couple of days ago. They will be included in RC2.
- In https://github.com/aspnet/EntityFramework/issues/4441 I have created an issue to track some other ideas on how to improve the experience for this scenario, e.g. if you need to create multiple instances of a `DbContext` in a request but you still want to use a nice pattern with DI. 

I will close this issue because I believe I have answered your questions but feel free to reopen if needed.

**Comment by divega** [comment]
@TrabacchinLuigi One more thing I realized while answering your question in SO: I forgot to ask you for more details about the circular reference. Maybe that is something you can avoid in other ways.

**Comment by TrabacchinLuigi** [comment]
Well avoiding circular references in queries I did in that method is easy the problem was finding out I got an user connected to the post I queried because that user is the logged one and also the creator of the post, maybe identity framework should use AsNoTracking() or I found a bug
  	 	
Also creating a new post gave me some troubles, after I added the entity with my userId only and did a SaveChanges I suddenly got the user related, and next step was to return it so I got a circular reference there too
  	 	
P. P. S. Another way of avoiding that is surely copy to new objects or dtos but I kinda dislike that if it is not really necessary... Maybe it's only me...
